Service/Microsoft.Health.DeIdentification.Local/InMemoryQueueClient.cs

```diff
@@ -1,37 +1,106 @@
+                }
+
+                if (jobInfo.Status == JobStatus.Running)
+                {
+                    jobInfo.CancelRequested = true;
+                }
+            }
+
+            return Task.CompletedTask;
         }
 
-        public Task CompleteJobAsync(JobInfo jobInfo, bool requestCancellationOnFailure, CancellationToken cancellationToken)
+        public async Task CompleteJobAsync(JobInfo jobInfo, bool requestCancellationOnFailure, CancellationToken cancellationToken)
         {
-            throw new NotImplementedException();
+            JobInfo jobInfoStore = jobInfos.FirstOrDefault(t => t.Id == jobInfo.Id);
+            jobInfoStore.Status = jobInfo.Status;
+            jobInfoStore.Result = jobInfo.Result;
+
+            if (requestCancellationOnFailure && jobInfo.Status == JobStatus.Failed)
+            {
+                await CancelJobByIdAsync(jobInfo.QueueType, jobInfo.Id, cancellationToken);
+            }
         }
 
         public Task<JobInfo> DequeueAsync(byte queueType, string worker, int heartbeatTimeoutSec, CancellationToken cancellationToken, long? jobId = null)
         {
-            throw new NotImplementedException();
+            JobInfo job = jobInfos.FirstOrDefault(t => t.Status == JobStatus.Created || (t.Status == JobStatus.Running && (DateTime.Now - t.HeartbeatDateTime) > TimeSpan.FromSeconds(heartbeatTimeoutSec)));
+            if (job != null)
+            {
+                job.Status = JobStatus.Running;
+                job.HeartbeatDateTime = DateTime.Now;
+            }
+
+            return Task.FromResult(job);
         }
 
-        public Task<IReadOnlyCollection<JobInfo>> DequeueJobsAsync(byte queueType, int numberOfJobsToDequeue, string worker, int heartbeatTimeoutSec, CancellationToken cancellationToken)
+        public async Task<IReadOnlyCollection<JobInfo>> DequeueJobsAsync(byte queueType, int numberOfJobsToDequeue, string worker, int heartbeatTimeoutSec, CancellationToken cancellationToken)
         {
-            throw new NotImplementedException();
+            var jobs = new List<JobInfo>();
+            while(numberOfJobsToDequeue > 0)
+            {
+                JobInfo job = jobInfos.FirstOrDefault(t => t.Status == JobStatus.Created || (t.Status == JobStatus.Running && (DateTime.Now - t.HeartbeatDateTime) > TimeSpan.FromSeconds(heartbeatTimeoutSec)));
+                if (job != null)
+                {
+                    job.Status = JobStatus.Running;
+                    job.HeartbeatDateTime = DateTime.Now;
+                }
+                jobs.Add(job);
+            }
+            return jobs;
         }
 
         public Task<IReadOnlyList<JobInfo>> EnqueueAsync(byte queueType, string[] definitions, long? groupId, bool forceOneActiveJobGroup, bool isCompleted, CancellationToken cancellationToken)
         {
-            throw new NotImplementedException();
+            var result = new List<JobInfo>();
+
+            long gId = groupId ?? largestId++;
```

Comment:
groupId should be the same for the jobs, we can set the default value to 0 instead of largestId.

Reply:
Ok, I will modify it